### PR TITLE
[parse] - Project cleanup and fix for Object.set/save

### PR DESCRIPTION
**Pull request by ridem**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes:
  - How set works: https://github.com/parse-community/Parse-SDK-JS/blob/70321e4c4080440f126c720ff9675e3e7f7d00b3/src/ParseObject.js#L699
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

Changes, in the order of the 4 commits:
- Align @types/parse with Definitely Typed's prettier rules: `npm run prettier -- --write types/parse/**`. See https://github.com/DefinitelyTyped/DefinitelyTyped/blob/394b6f5ceedd3acaea0d183f8d2c9b7117c8b3f3/README.md#common-mistakes
- Various project cleanup fixes
  - Removes the `ts4.0` folder. Tests for the default types are passing from 3.7 to 4.3, and a split at ts4.0 is quire rare in the rest of this repo (even @types/node doesn't have one).
  - Set up the project so that `no-self-import` isn't needed. This means `node.d.ts` imports `./index` instead of `parse`.
- Re-align `ts3.6` with the last changes from the default types folder.
- Fix for `Parse.Object` `set` and `save` methods. The tricks in use don't appear to be fixing any issue but make code completion difficult. They also break in some cases

### CODE REVIEW

File: types/parse/index.d.ts [hunk 1]
```diff
@@ -457,25 +457,9 @@ declare global {
             remove: this["add"];
             removeAll: this["addAll"];
             revert(...keys: Array<Extract<keyof (T & CommonAttributes), string>>): void;
-            save<K extends Extract<keyof T, string>>(
-                attrs?:
-                    | (((x: T) => void) extends (x: Attributes) => void
```
Comment (ridem) [inline types/parse/index.d.ts:462]: This is true:
```ts
((x: T) => void) extends (x: Attributes) => void
```
only when `T` has no optional key.

**Review by rdhelms — CHANGES_REQUESTED** [review]
@ridem What's your use case that led you to make this PR? The change as you have it would break existing functionality. 
~~Also, it looks like you had some autosave lint config applied here, which changed all the `'` -> `"`.~~ 
Interesting...I didn't know DefinitelyTyped updated their single/double quote preference [a couple months ago](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/51436)

**Comment by ridem** [comment]
Thanks for the quick answer, @rdhelms.

> What's your use case that led you to make this PR? 

Several reasons:

1. Keep IntelliSense working with `set()` and `save()`

Not having autocompletion suggestions on things like: `myObject.set({})` is extremely annoying. 

2. Being able to work with `Partial<Attributes>` in the codebase.

Basically being able to do things like this:
```ts
type BookAttributes = {
    author: string; title: string;
}

class Book extends Parse.Object<BookAttributes> {
    constructor(options?: any) {
        super("Book", options);
    }

    setTolkienBook(attrs: Partial<BookAttributes>) {
        this.set({
            ...attrs,
            author: "Tolkien"
        })
    }
}
```

3. Set up the project so that it's less intimidating for future PRs

This is a warmup for which I read the `README` and fixed a few things to get `@types/parse` to be closer to the conventions. Removing an unnecessary folder and aligning the formatting all went in that direction. I intend on adding support for dot notation queries with a `typesVersions` on ts4.1.

---

About `undefined` values:

Thanks for clarifying what the intent was with this. I've added a commit that includes:
- Tests for when `undefined` is a value, in various contexts. Notably, the previous types for `set` and `save` weren't reliable when all keys of the `Attributes` were already optional (tests using `objTypedOptional` were failing)
- Updated definitions for `set` and `save` that forbid `undefined` while still making IntelliSense work

For my use case n°2, I understand that this isn't possible if you do rely on forbidding `undefined` values in your setup, so I'm not going forward with this. I don't have a full picture over how many people rely on this guard versus how many people are annoyed that they can't simply pass a `Partial` of their attributes to `set` and `save`.

AFAIK Parse isn't meant to throw when you pass an `undefined` value, the behaviour is just a bit weird:
- The object's attribute is set locally, but it's a no-op on the server
- There might still be a bug for when called server-side with `directAccess: true`, but this isn't intended.

There are other places which would need guards to forbid `undefined` values, but are tricky to implement, notably:
```ts
class Test extends Parse.Object<{title?: string;}> {}
new Test("Test", { title: undefined })
```
or
```ts
function testSet( objUntyped: Parse.Object) {
    // $ExpectType false | Object<Attributes>
    objUntyped.set({ propA: undefined });
}
```

Is it the types job to forbid these `undefined` values in a few cases (and not others) at the expense of other useful uses?

https://github.com/microsoft/TypeScript/issues/43805 is an interesting read about how the issue might be tackled in future releases.
 
---

To sum up what the PR is doing:
- Remove an unnecessary `typesVersions` folder (ts4.0)
- Remove a tslint rule override which was easy to get rid of
- Apply DefinitelyTyped's styling guidelines to the project (`npm run prettier -- --write types/parse/**`)
- Put back IntelliSense on `.set({})` and `.save({})` 
- [No change] `undefined` is still forbidden as a value in `set` and `save` when `Attributes` are specified. Added tests and comments to make that intent clearer

**Comment by rdhelms** [comment]
@ridem Thanks for the really useful explanation and summary. I'm testing your changes locally a bit before approving, but that all seems very promising. 

We hadn't noticed any issues with Intellisense in our codebase, because rather than using classes that extend Parse Object classes, we just create and use the Parse Objects directly, and `.set()` and `.save()` seem to work fine in those cases, even with `Partial` attributes. As an example (using `@types/parse@2.18.3`):
image

But it seems like you've improved the handling for cases like extended classes, so that seems great 👍🏻

For using `undefined`, our intent when first implementing generically typed Parse Objects wasn't to completely forbid passing `undefined` in all cases, it was just to make the passed values match the user's "schema" (aka the types of the Parse Object's attributes). In other words, if the attributes object has a property typed using `?` or `undefined`, then the intent is to allow `new Parse.Object()`, `.set()`, and `.save()` to pass `undefined` for that property. But if the property is a required property, then the intent was to prevent accidentally setting that property's value to be `undefined`. So as an example, for your case of the `Book` class with `setTolkienBook`, I would expect that attempting to call `this.set()` with anything other than an object matching `BookAttributes` _should_ fail.
```ts
class Book extends Parse.Object<BookAttributes> {
    constructor (options: BookAttributes) {
        super('Book', options)
    }

    setTolkienBook () {
        this.set({
            author: undefined,    // Errors, as expected
        })
    }
}
```

But, as you've rightfully pointed out, the existing types allow `this.set({})` there, which to me is a bug.
```ts
class Book extends Parse.Object<BookAttributes> {
    constructor (options: BookAttributes) {
        super('Book', options)
    }

    setTolkienBook () {
        this.set({})    // Should not work
    }
}
```

I will also say that we've had a relatively limited scope of use cases and documented examples of how people are using this package, so I think it would be interesting to compare notes some more to make sure that we're taking as many cases, patterns, and architectures into account as possible.

@RaschidJFR or others, would you care to share how using this package (particularly the generically-typed `Parse.Object` with `.save` and `.set`) has been going for you? Have you noticed any other particular frustrations related to Intellisense or objects with `undefined` properties?

File: types/parse/parse-tests.ts [hunk 2]
```diff
@@ -1325,145 +1329,192 @@ function testObject() {
...
+
+        // $ExpectError
+        await objTyped.save("example", undefined);
+
+        // $ExpectError
+        await objTypedOptional.save({ example: undefined });
 
         // $ExpectError
-        await objTyped.save('wrongProp', true);
+        await objTypedOptional.save("example", undefined);
```
Comment (rdhelms) [inline types/parse/parse-tests.ts:1464]: Ah...I misinterpreted what you were saying in your other comment...so actually in these cases, since the attributes type explicitly types the properties as `undefined`, we would have allowed setting these. I do see how it could be beneficial to prevent ever setting a value to be `undefined` in order to force/guide users more towards `.unset`...but I do think that may be a bit heavy-handed.

More and more, I think the core issue for your case anyway is some difference between the behavior of `class Thing extends Parse.Object<SomeAttributes>` and `new Parse.Object<SomeAttributes>('Thing')`. It seems like the extended class is losing some TS context somehow...

**Review by rdhelms — CHANGES_REQUESTED** [review]
@ridem Ok - I think maybe I've come up with a solution to help with your Intellisense and extended class use case, while maintaining the existing behavior of the package thus far:
```ts
            save<K extends Extract<keyof T, string>>(
                attrs?: Pick<T, K> | T | null,
                options?: Object.SaveOptions,
            ): Promise<this>;
            save<K extends Extract<keyof T, string>>(key: K, value: T[K], options?: Object.SaveOptions): Promise<this>;
            set<K extends Extract<keyof T, string>>(
                attrs: Pick<T, K> | T | null,
                options?: Object.SetOptions,
            ): this | false;
            set<K extends Extract<keyof T, string>>(key: K, value: T[K], options?: Object.SetOptions): this | false;
```

I think this is possible now partly due to updates by the TS compiler since the time when these types were first written. The `| T` part of the union type is what particularly helps Intellisense, and that idea came from react's types for `setState` here:
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react/index.d.ts#L482

If that seems good to you, it would be great if you could adjust that but keep all the other improvements with linting, etc. And I would love to get a future PR figured out for handling dot notation with template literal types 😄

File: types/parse/index.d.ts [hunk 3]
```diff
@@ -29,7 +29,14 @@
...
 /// <reference path="react-native.d.ts" />
 
-import { EventEmitter } from 'events';
+import { EventEmitter } from "events";
+
+/**
+ * Definitions for save and set forbid undefined values while keeping Intellisense working
+ * See https://github.com/DefinitelyTyped/DefinitelyTyped/blob/3bdadbf9583c2335197c7e999b9a30880e055f62/types/react/index.d.ts#L482
+ */
+type SafeAttributes<T, K extends keyof T> = Pick<Required<T>, K> | Required<T>;
```
Comment (ridem) [inline types/parse/index.d.ts:38]: @rdhelms Sorry I wasn't clearer in my explanation then, this is also the approach I took, with a link to the React types

Comment (ridem) [reply types/parse/index.d.ts:38]: Except we need to use `Required` in case some attributes are "optional"

Comment (rdhelms) [reply types/parse/index.d.ts:38]: Do we really want to start preventing setting undefined in all cases though? That's not what the current behavior is, and it's not what the original intent was, and you mentioned that it may not be the job of the types package to enforce that.

File: types/parse/parse-tests.ts (hunk 2 above)
Comment (rdhelms) [reply types/parse/parse-tests.ts:1464]: I would vote that we support these test cases:
```ts
        // $ExpectError
        await objTyped.save({ example: undefined });

        // $ExpectError
        await objTyped.save("example", undefined);

        // $ExpectType Object<OptionalObjectAttributes>
        await objTypedOptional.save({ example: undefined });

        // $ExpectType Object<OptionalObjectAttributes>
        await objTypedOptional.save("example", undefined);
```

Which, by removing the `Required<>` from your `save()` and `set()` types, I believe should all pass and also still support Intellisense for all kinds of attribute types, partial or not.

File: types/parse/index.d.ts (hunk 3 above)
Comment (rdhelms) [reply types/parse/index.d.ts:38]: If you're on board with allowing `undefined` for cases where the attributes are specifically typed as either `Partial` or properties with `undefined`, I would vote to remove these helper types and just do `Pick<T, K> | T` at the `save` and `set` definitions.

File: types/parse/index.d.ts (hunk 1 above)
Comment (rdhelms) [reply types/parse/index.d.ts:462]: To clarify here: the only reason why this case _should_ error (in my opinion) is because the attributes had been typed without allowing `undefined`. For the case where the attributes ARE typed either as `Partial` or with properties typed as `undefined`, I'm fine allowing those to be set as `undefined`.

My problem was just with the `Partial<T>`, but I think `Pick<T, K> | T` seems to work great.

**Review by rdhelms — CHANGES_REQUESTED** [review]

**Comment by ridem** [comment]
Ok @rdhelms thanks for clarifying the test cases that we should support. I had intentionally added `objTypedOptional` to raise this: it was hard to grasp what the trick in use before was trying to achieve.

My personal preference is still what I had initially published in the PR: set and save simply being `Partial<T>`, given that this is what the actual implementation of set and save suggest. Passing `undefined` is never a good idea no matter the types, but I agree it's not the type's job to check for this.

That being said, I think we reached a good compromise and I'll update according to what you've suggested.

To follow up on the discussion around Parse's behaviour around defined keys with nullable values, here's a sample of its runtime behaviour (example uses save, but set is similar)
```ts
const mk = { useMasterKey: true };
class Organization extends Parse.Object {
  constructor(opts) {
    super("Organization", opts);
  }
}
Parse.Object.registerSubclass("Organization", Organization);

const orgQuery = new Parse.Query(Organization);
const org = await orgQuery.first(mk);
org.get("name"); // "MyOrg"

// Saving "null"

await org.save({ name: null }, mk);
org.get("name"); // null

await org.fetch(mk);
org.get("name"); // undefined

// Saving a string

await org.save({ name: "ACME" }, mk);
org.get("name"); // "ACME"

await org.fetch(mk);
org.get("name"); // "ACME"

// Saving "undefined"

await org.save({ name: undefined }, mk);
org.get("name"); // undefined

await org.fetch(mk);
org.get("name"); // "ACME"

// Saving an empty object

await org.save({}, mk);
org.get("name"); // "ACME"

await org.fetch(mk);
org.get("name"); // "ACME"
```

Main takeaway is that it's totally ok to do `set({})` but `set({key: undefined})` is never a good idea.

File: types/parse/index.d.ts (hunk 3 above)
Comment (ridem) [reply types/parse/index.d.ts]: Agree, sounds like a good compromise. I pinged you here because I felt bad I didn't properly explain that the `Pick<T, K> | T` trick was already part of the PR, that's what I meant when I said "keep IntelliSense working"

**Review by rdhelms — APPROVED** [review]
Thanks @ridem - I think this is great. It's some much needed cleanup, improved transparency, and a great fix for Intellisense on `save` and `set`.

I really appreciate the compromise we came to, because I know for my company's particular use we highly value how TS will error if we try to set something as `undefined` - we see that as a bug and never want to allow it in our own codebase. But since not everyone would want to be that strict, I think having the ability to opt out of those errors by using `Partial` or `| undefined` on an attribute is really valuable. And for those wanting even looser typings, we still have the parameter-less flavor of `Parse.Object()` which lets people set and save whatever attributes they want.

It's incredibly ironic that we independently found and cited the exact same line of react's types...initially when I saw your change with `SafeAttributes`, I was only focused on the `Required` part and didn't even think about the `| T` part. And then after my own research attempts, I just ended up at the exact thing that you had already found 😅

Let us know how the support for dot notation goes - we would love to get support for that, and are happy to help out if the effort becomes problematic.

**Comment by rdhelms** [comment]
Also, it seems like the CI builds are failing for something unrelated to parse - not sure what that's about. Looks like maybe it's from `rxjs`
https://github.com/DefinitelyTyped/DefinitelyTyped/commit/7df6984e55bc8c26589802d9d21497cc31dd46d3

**Comment by rdhelms** [comment]
@sandersn It looks like our package's build is failing with [errors related to rxjs](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/52592/checks?check_run_id=2477640528)...is it possible that we need to make updates related to your recent [commit](https://github.com/DefinitelyTyped/DefinitelyTyped/commit/7df6984e55bc8c26589802d9d21497cc31dd46d3)?

**Comment by sandersn** [comment]
Yep, you'll need to merge from master to get rid of those errors.

**Comment by ridem** [comment]
@rdhelms 
For dot notation, I've published a working concept with test cases in my fork, under a  parse/ branch if you want to have a look at it.
The main issue around it is choosing a solution for when a class could have a pointer to itself. Either this is fixed by setting a limit on recursion with a nice iterator, or with some smart detection forbidding a class including itself

**Review by rdhelms — APPROVED** [review]

**Comment by ridem** [comment]
Ready to merge

**Comment by lamualfa** [comment]
@ridem @rdhelms @sandersn Thank you very much.

But i have some problem when using ES 6 Minimized version as mentioned in official [repo](https://github.com/parse-community/Parse-SDK-JS#using-parse-on-different-platforms):

```js
import Parse from 'parse/dist/parse.min.js';
```

When using the method above, the autocomplete will not showing. How to show the autocomplete even i use ES 6 Minimized version?
